SHA3.cs

```diff
@@ -0,0 +1,637 @@
+    public static ValueTask<byte[]> HashDataAsync(byte[] key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<byte[]> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<int> HashDataAsync(
+        ReadOnlyMemory<byte> key,
+        Stream source,
+        Memory<byte> destination,
+        CancellationToken cancellationToken = default);
+}
+
+public abstract partial class Shake : IDisposable {
+    internal Shake(); // Will not be subclassable by developers.
+
+    public void AppendData(byte[] data);
+    public void AppendData(ReadOnlySpan<byte> data);
+    protected abstract void AppendDataCore(ReadOnlySpan<byte> data);
+
+    public byte[] GetCurrentHash(int outputLength);
+    public void GetCurrentHash(Span<byte> destination);
+    protected abstract void GetCurrentHashCore(Span<byte> destination);
+
+    public byte[] GetHashAndReset(int outputLength);
+    public void GetHashAndReset(Span<byte> destination);
+    protected abstract void GetHashAndResetCore(Span<byte> destination);
+
+    public void Dispose();
+    protected virtual void Dispose(bool disposing);
+}
+
+[UnsupportedOSPlatform("ios")]
+[UnsupportedOSPlatform("tvos")]
+[UnsupportedOSPlatform("android")]
+[UnsupportedOSPlatform("browser")]
+[UnsupportedOSPlatform("osx")]
+public sealed partial class Shake128 : Shake {
+
+    public Shake128();
+
+    [UnsupportedOSPlatformGuard("ios")]
+    [UnsupportedOSPlatformGuard("tvos")]
+    [UnsupportedOSPlatformGuard("android")]
+    [UnsupportedOSPlatformGuard("browser")]
+    [UnsupportedOSPlatformGuard("osx")]
+    public static bool IsSupported { get; }
+
+    public static byte[] HashData(byte[] source, int outputLength);
+    public static byte[] HashData(ReadOnlySpan<byte> source, int outputLength);
+    public static void HashData(ReadOnlySpan<byte> source, Span<byte> destination);
+
+    public static byte[] HashData(Stream source, int outputLength);
+    public static void HashData(Stream source, Span<byte> destination);
+    public static ValueTask HashDataAsync(Stream source, Memory<byte> destination, CancellationToken cancellationToken = default);
+    public static ValueTask<byte[]> HashDataAsync(Stream source, int outputLength, CancellationToken cancellationToken = default);
+}
+
+[UnsupportedOSPlatform("ios")]
+[UnsupportedOSPlatform("tvos")]
+[UnsupportedOSPlatform("android")]
+[UnsupportedOSPlatform("browser")]
+[UnsupportedOSPlatform("osx")]
+public sealed partial class Shake256 : Shake {
```

Comment:
I'm of two minds on sealing SHAKE128 and 256.  If they were supported by all of our platforms, it'd be easy... sealing it closes off the "I extended it with a wrong answer" problems, and the "I'm trying to provide a fallback implementation for my platform" doesn't apply.

I guess the fallback doesn't apply anyways for the static one-shots, which is probably going to be the more common call pattern, which then makes the BYO-impl not really relevant, so I guess I've talked myself out of being of two minds and maybe think it's just right to seal it.

Reply:
We sealed off ChaCha20Poly1305 and no one complained in the years that it has been in, also with somewhat limited OS support.

There is the base Shake class, which I think provides some reasonable extension point for someone implementing their own SHAKE, if they want. The trick would be making sure consuming APIs take a `Shake` instead of a `Shake128`, but, all in all I also just agree most folks will reach for the one-shots and the extendability is not interesting to people.

Reply:
> There is the base Shake class, which I think provides some reasonable extension point for someone implementing their own SHAKE, if they want.

Well, that type you decorated with an `internal` ctor, so it can't be extended outside the S.S.Cryptography assembly.

But, yeah, seal-until-evidence-says-otherwise seems good here, given the statics.

Reply:
Ah, ha, yeah... I am very distracted today with $COMPANYINTERNALS.